db/migrations/20200305202405_create_get_or_create_header_function.sql

```diff
@@ -0,0 +1,43 @@
+-- +goose Up
+-- SQL in this section is executed when the migration is applied.
+
+-- +goose StatementBegin
+CREATE OR REPLACE FUNCTION public.get_or_create_header(block_number BIGINT, hash VARCHAR(66), raw JSONB, block_timestamp NUMERIC, eth_node_id INTEGER) RETURNS INTEGER AS
+$$
+DECLARE
+    matching_header_id INTEGER := (
+        SELECT id
+        FROM public.headers
+        WHERE headers.block_number = get_or_create_header.block_number
+        AND headers.hash = get_or_create_header.hash
+        );
+    nonmatching_header_id INTEGER := (
+        SELECT id
+        FROM public.headers
+        WHERE headers.block_number = get_or_create_header.block_number
+        AND headers.hash != get_or_create_header.hash
+        );
+    inserted_header_id INTEGER;
+BEGIN
+    IF matching_header_id != 0 THEN
+        RETURN matching_header_id;
+    END IF;
+
+    IF nonmatching_header_id != 0 THEN
```

Comment:
Previewing vdb-1181 - I think we can add a predicate that the non-matching header block number needs to be within 15 of the max block in the db in order to be replaced (so that a forked node stuck on an older block isn't deleting/replacing very old blocks that are probably canonical). Will need to think further about what we want to do with those headers that appear to be coming from a fork (e.g. insert them anyway? drop them silently? raise an exception? 🤔)